test/e2e/installed_package_test.go

```diff
@@ -143,3 +144,102 @@ stringData:
+	env := BuildEnv(t)
+	logger := Logger{}
+	kapp := Kapp{t, env.Namespace, logger}
+	kubectl := Kubectl{t, env.Namespace, logger}
+	sas := ServiceAccounts{env.Namespace}
+	name := "instl-pkg-test-fail"
+
+	// contents of this repo (k8slt/installedpackage-fail)
+	// under test/e2e/assets/installedpackage-fail-test
+	installPkgYaml := fmt.Sprintf(`---
+apiVersion: install.package.carvel.dev/v1alpha1
+kind: PackageRepository
+metadata:
+  name: fail-repo
+  # cluster scoped
+spec:
+  fetch:
+    image:
+      url: k8slt/installedpackage-fail@sha256:d17671179e8304f6cb81a4e989d1724c795b506d07669cd8c375ab4e5e012d47
+---
+apiVersion: install.package.carvel.dev/v1alpha1
+kind: InstalledPackage
+metadata:
+  name: %s
+  namespace: %s
+  annotations:
+    kapp.k14s.io/change-group: kappctrl-e2e.k14s.io/installedpackages
+spec:
+  serviceAccountName: kappctrl-e2e-ns-sa
+  packageRef:
+    publicName: pkg.fail.carvel.dev
+    version: 1.0.0
+  values:
+  - secretRef:
+      name: pkg-demo-values
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: pkg-demo-values
+stringData:
+  values.yml: |
+    #@data/values
+    ---
+    hello_msg: "hi"
+`, name, env.Namespace) + sas.ForNamespaceYAML()
+
+	cleanUp := func() {
+		// Delete App with kubectl first since kapp
+		// deletes ServiceAccount before App
+		kubectl.RunWithOpts([]string{"delete", "apps/" + name}, RunOpts{AllowError: true})
+		kapp.Run([]string{"delete", "-a", name})
+	}
+	cleanUp()
+	defer cleanUp()
+
+	// Create Repo, InstalledPackage, and App from YAML
+	kapp.RunWithOpts([]string{"deploy", "-a", name, "-f", "-"}, RunOpts{StdinReader: strings.NewReader(installPkgYaml)})
+
+	expectedStatus := v1alpha12.InstalledPackageStatus{
+		GenericStatus: v1alpha1.GenericStatus{
+			Conditions: []v1alpha1.AppCondition{{
+				Type:    v1alpha1.ReconcileFailed,
+				Status:  corev1.ConditionTrue,
+				Message: "Error (see .status.usefulErrorMessage for details)",
+			}},
+			ObservedGeneration:  1,
+			FriendlyDescription: "Reconcile failed: Error (see .status.usefulErrorMessage for details)",
+		},
+		Version: "1.0.0",
+	}
+
+	// wait for status to update for InstalledPackage
+	var cr v1alpha12.InstalledPackage
+	retryFunc := func() error {
+		out := kubectl.Run([]string{"get", fmt.Sprintf("ipkg/%s", name), "-o", "yaml"})
+		err := yaml.Unmarshal([]byte(out), &cr)
+		if err != nil {
+			return fmt.Errorf("Failed to unmarshal: %s", err)
+		}
+
+		if !strings.Contains(cr.Status.UsefulErrorMessage, "kapp: Error") {
+			return fmt.Errorf("\nExpected useful error message to contain deploy error\nGot:\n%s", cr.Status.UsefulErrorMessage)
+		}
+
+		return err
+	}
+	err := retry(30*time.Second, retryFunc)
+	if err != nil {
+		t.Fatalf("Expected error from InstalledPackage %s: %v", name, err)
+	}
+	// Remove error message since asserted on already
+	cr.Status.UsefulErrorMessage = ""
+
+	if !reflect.DeepEqual(expectedStatus, cr.Status) {
```

Comment:
can we consolidate this check with the one in the retry? Useful error message is a field on status, so can we just assert status is equal in there instead of specifically asserting on useful error message? Alternatively, if we want it separate for explicitness of test, maybe we just assert on UsefulErrorMessage and remove the deep equals check since the test scope seems to be focused on just UsefulErrorMessage

Reply:
I guess what we can do is also check that the `FriendlyDescription` updates in the retry. I think that is the other main property we care about.

Reply:
Updated